### ISSUE TITLE
Nested component in Portal not receiving state values of React Context

### ISSUE THREAD

**Issue by ismailcodeboi** [callstack/react-native-paper#3880]
### Current behaviour
When using Portal component you are not given the context of the state in the wrapping/global React Context Provider

### Expected behaviour
You should be able to access the state values of the Context Provider

### How to reproduce?
- create a new expo app
- install react-native-paper 
- create a context using createContext from react and give it an initial state/default value
- wrap the entire app in your named context.Provider
- create any component (text, view, snackbar etc.)
- using the Portal component wrap the created component
- inside the created component call the useContext() with your created context as the parameter
- expect the state to be defined/a value, but it is undefined

https://snack.expo.dev/184285Vsi - here is a simple snack to what the issue is, with some logs

### Preview
The Context and Provider
![image](https://github.com/callstack/react-native-paper/assets/41947220/5933da12-2b5a-47bd-9a25-25865a78395f)

Main App wrapped in Provider
![image](https://github.com/callstack/react-native-paper/assets/41947220/99e54a63-a8e9-4328-8b0e-fa7a32a800cd)

Portal Component
![image](https://github.com/callstack/react-native-paper/assets/41947220/9fcec0b7-cade-495a-827b-d100b6ea7669)

nested component of Portal
![image](https://github.com/callstack/react-native-paper/assets/41947220/448ca2be-8994-42d8-aa35-3a96f030e232)

Log of initial values outside of Portal
![image](https://github.com/callstack/react-native-paper/assets/41947220/20138a2e-4135-4bbc-bd71-aded9b800829)

We expect these values to be the same within the portal

However as seen the in screenshot of the logs below it is undefined: 
![image](https://github.com/callstack/react-native-paper/assets/41947220/8428389e-e1c9-4966-8acf-a2437ae09499)

### What have you tried so far?
To fix this I have passed in the state values as props to the nested component - this seems to work, however it is defeating the purpose of using Context Providers of react

### Your Environment

| software                      | version
| -----------------------  | -------
| react-native                | 0.71.7
| react-native-paper     | ^5.7.0
| node                           | 16.19.0
| npm                            | 8.19.3
| expo sdk                     | ^48.0.11

**Comment by JPStrydom** [comment]
Thanks for raising this issue! We're experiencing the same issue - took us days to debug.

**Comment by JPStrydom** [comment]
> Try flipping the positions of `<StateProvider>` and `<Provider>`

But then we won't get access to the theme in our `StateProvider` will we? Which might be a bit of a pain as I can imagine things like error boundaries, notifications, snackbars, etc. might be needed outside the `StateProvider`

**Comment by JPStrydom** [comment]
I still fell like this is breaking the React rule. If something is inside a React context, it should have access to that context - where as that is not the case with the RNP `Portal` component. 

And the fix of simply moving the state context outside of the theme context seems like a hack, as now we'll not have access to the theme inside the state provider.

**Comment by andrewtran10** [comment]
Anybody have a reason as to why this occurs?

**Comment by jeferson-sb** [comment]
@ismailcodeboi I tinked a bit with your example and seems that the `<Portal>` component will always render outside of the parent tree, while `<Portal.Host>` is the one that renders the modal alongside others components in the tree and can access the _StateContext_.

Not really sure if it should behave this way, but it how defined in the render function of the Portal component:
(see line 59 where it renders children)

https://github.com/callstack/react-native-paper/blob/1ce31113cd67e7aa450c72d6f86cd4a20c8157a1/src/components/Portal/Portal.tsx#L44-L68

**Comment by JPStrydom** [comment]
One big problem that we currently have with this issue, is that our modals cannot read the navigation state or use the navigation hooks. We're using [React Navigation](https://reactnavigation.org/) for our navigation, which requires routes at the root level of it context wrapper - meaning that we can't have our theme context nor our own state context inside our navigation context. This means that we won't be able to access these contexts inside our modals, which is quite frustrating as we can't have any navigation functionality on our modals now.

**Comment by emanoelqueiroz** [comment]
I'm facing the same problem thank you for open this issue. You guys have any elegant workaround for this?

**Comment by JPStrydom** [comment]
> I'm facing the same problem thank you for open this issue. You guys have any elegant workaround for this?

Unfortunately not an elegant solution, but we ended up wrapping our RNP `ThemeProvider` in our state providers. This did require  a bit of rework of our application. We had to move our theme and state consumers inside the theme provider (which is inside our state providers). This then allowed all `Portal` components to have access to the relevant context.

This is still not an ideal solution, as our `NavigationProvider` (from React Navigation) cannot be outside the `ThemeProvider` due to the way it works. This means that modal components cannot consume the navigation context (navigation state and navigation functions).

**Comment by emanoelqueiroz** [comment]
Yeah, I ended up doing the same and facing the same problem of not being able to use anything from React Navigation.

I'll try some workarounds (like I've being doing) and when they fix, maybe, I'll change this "temporary code", but thank you anyway buddy

**Comment by danb4r** [comment]
> I still fell like this is breaking the React rule. If something is inside a React context, it should have access to that context - where as that is not the case with the RNP `Portal` component.
> 
> And the fix of simply moving the state context outside of the theme context seems like a hack, as now we'll not have access to the theme inside the state provider.

Totally agree. Also took hours to find out. And the same happens with React Navigation context as reported in #4022 .

**Comment by szado** [comment]
I'm also facing the same problem.

**Comment by Bi0max** [comment]
Same problem

**Comment by rveltonCL** [comment]
+1, facing the same issue. Is there an update for this?

**Comment by JacobDel** [comment]
Any idea for how to attract some attention from RN paper devs? This issue seems abandoned. Would be petty to make a MR with something they don't approve

**Comment by artiphishle** [comment]
Could anyone explain, why my fix below (comments in code) actually works (I'm combining with`reduceRight`)

```ts
export function combineProviders(
  ...providers: ComponentType<IProviderProps>[]
) {
  return function CombinedProviders({ children, ...props }: IProviderProps) {
    return providers.reduceRight(
      (acc, Provider) => <Provider {...props}>{acc}</Provider>,
      children
    );
  };
}

export const AnkhProviders = combineProviders(
  ConfigProvider,
  // Before: PaperThemeProvider here
  LanguageProvider,
  AuthProvider,
  PaperThemeProvider // After: Moved to get AuthProvider's context
);
```
